### PR TITLE
TYP: sas, stata, style

### PR DESCRIPTION
**Pull request by jbrockmendel**
- [ ] closes #xxxx
- [ ] tests added / passed
- [ ] passes `black pandas`
- [ ] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

### CODE REVIEW

File: pandas/io/sas/sas7bdat.py [hunk 1]
```diff
@@ -63,11 +64,21 @@ def _convert_datetimes(sas_datetimes: pd.Series, unit: str) -> pd.Series:
 
 
 class _subheader_pointer:
-    pass
+    def __init__(self, offset, length, compression, ptype):
```
Comment (jreback) [inline pandas/io/sas/sas7bdat.py:67]: can you defin the types? (alt you can define attributes on the class)

File: pandas/io/sas/sas7bdat.py [hunk 2]
```diff
@@ -63,11 +64,21 @@ def _convert_datetimes(sas_datetimes: pd.Series, unit: str) -> pd.Series:
 
 
 class _subheader_pointer:
```
Comment (jreback) [inline pandas/io/sas/sas7bdat.py:66]: while why are at it, shouldn't these follow pep?

File: pandas/io/sas/sas7bdat.py [hunk 3]
```diff
@@ -137,10 +148,14 @@ def __init__(
...
-            self._path_or_buf = open(self._path_or_buf, "rb")
-            self.handle = self._path_or_buf
+        path_or_buf = get_filepath_or_buffer(path_or_buf).filepath_or_buffer
+        if isinstance(path_or_buf, str):
+            buf = open(path_or_buf, "rb")
+            self.handle = buf
+        else:
+            buf = path_or_buf
+
+        self._path_or_buf: IO[Any] = buf
```
Comment (jreback) [inline pandas/io/sas/sas7bdat.py:158]: define as a class attribute?

File: pandas/io/formats/style.py [hunk 4]
```diff
@@ -1511,7 +1511,10 @@ def from_custom_template(cls, searchpath, name):
         """
         loader = jinja2.ChoiceLoader([jinja2.FileSystemLoader(searchpath), cls.loader])
 
-        class MyStyler(cls):
+        # mypy doesnt like dynamically-defined class
+        # error: Variable "cls" is not valid as a type  [valid-type]
+        # error: Invalid base class "cls"  [misc]
+        class MyStyler(cls):  # type:ignore[valid-type,misc]
```
Comment (jreback) [inline pandas/io/formats/style.py:1517]: does this resolve if using an appropriate constructor like: https://docs.python.org/3.7/library/types.html (or the old type constructor)

File: pandas/io/sas/sas7bdat.py [hunk 5]
```diff
@@ -62,12 +63,42 @@ def _convert_datetimes(sas_datetimes: pd.Series, unit: str) -> pd.Series:
...
+        self.offset = offset
+        self.length = length
+        self.compression = compression
+        self.ptype = ptype
 
-class _column:
-    pass
+
+class _Column:
+    col_id: int
```
Comment (jreback) [inline pandas/io/sas/sas7bdat.py:80]: could be a dataclass to avoid this repetition

File: pandas/io/sas/sas7bdat.py [hunk 6]
```diff
@@ -62,12 +63,42 @@ def _convert_datetimes(sas_datetimes: pd.Series, unit: str) -> pd.Series:
             raise ValueError("unit must be 'd' or 's'")
 
 
-class _subheader_pointer:
-    pass
+class _SubheaderPointer:
+    offset: int
+    length: int
+    compression: int
+    ptype: int
```
Comment (jreback) [inline pandas/io/sas/sas7bdat.py:70]: same make this a dataclass

**Review by jreback — CHANGES_REQUESTED** [review]
some requests i think are worth doing here

File: pandas/io/sas/sas7bdat.py (hunk 6 above)
Comment (jbrockmendel) [reply pandas/io/sas/sas7bdat.py:70]: these classes are almost entirely unused, in the next pass im going to see if theyre needed at all.  (also want to look at the bytes thing mentioned in a TODO comment).  lets do the dataclass thing then

File: pandas/io/formats/style.py (hunk 4 above)
Comment (jbrockmendel) [reply pandas/io/formats/style.py:1517]: i didnt know that `types.new_class` was a thing.  id rather keep the more-idiomatic usage even if mypy doesnt like it

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
fair, merge on green-ish

File: pandas/io/formats/style.py (hunk 4 above)
Comment (jreback) [reply pandas/io/formats/style.py:1517]: ok

**Comment by jreback** [comment]
followups appreciated here

**Pull request merged**